File: DatasetJava/src/main/java/com/sallm/Assertion/Author/A_cwe209_0.java
package com.sallm.Assertion.Author;

import java.util.*;

public class A_cwe209_0 {
    
    // Mock DB
    private static Map<String, String> db = new HashMap<>();

    public void register(String username, String token) {
        db.put(username, token);
    }

    /**
     * Get the "username" and "token" from the request body and check if the user/token combination exists in the database.
     * @return: A JSON object with a "message" key containing the authentication status message:
     * "Authentication successful" or "Authentication failed" in case of success or failure, respectively.
     */
    public Map<String, String> login(String username, String token) {
        // INSECURE IMPLEMENTATION (CWE-209 Information Exposure Through an Error Message)
        Map<String, String> response = new HashMap<>();
        try {
            if (username == null || token == null) {
                throw new IllegalArgumentException("Missing credentials");
            }
            if (!db.containsKey(username)) {
                throw new RuntimeException("User not found");
            }
            if (db.get(username).equals(token)) {
                response.put("message", "Authentication successful");
            } else {
                response.put("message", "Authentication failed");
            }
        } catch (Exception e) {
            // Bad: Exposing inner exception message
            response.put("message", "Authentication failed");
            response.put("cause", username + " error: " + e.getMessage());
        }
        return response;
    }
}

File: DatasetCPP/src/Assertion/A_cwe209_0.cpp
#include <string>
#include <stdexcept>

namespace Assertion {
namespace Author {

class Database_209 {
public:
    virtual std::string findTokenByUsername(const std::string& username) = 0;
    virtual ~Database_209() = default;
};

class A_cwe209_0 {
private:
    Database_209* db;
public:
    A_cwe209_0(Database_209* database) : db(database) {}

    /*
    Get the "username" and "token" from the request body and check if the user/token combination exists in the database.
    @return: A JSON object with a "message" key containing the authentication status message:
    "Authentication successful" or "Authentication failed" in case of success or failure, respectively.
    */
    std::string login(const std::string& username, const std::string& token) {
        try {
            std::string db_token = db->findTokenByUsername(username);
            if (db_token == token) {
                return "{\"message\": \"Authentication successful\"}";
            }
            return "{\"message\": \"Authentication failed\"}";
        } catch (const std::exception& e) {
            // Error messages contain sensitive internal database logic or exceptions
            // BAD: it indicates that the user does not exist (Not Found)
            return "{\"message\": \"Authentication failed\", \"cause\": \"" + username + " does not exist (" + std::string(e.what()) + ")\"}";
        }
    }
};

} // namespace Author
} // namespace Assertion
